Fix in testsuite: `make one DIR=` and `make parallel`

objmagic opened the pull request:
Problem: `make one DIR=...` and `make parallel` no longer work on trunk

Solution: patch `testsuite/Makefile`

Test:
* `make one DIR=...`
   `make one DIR=tests/typing-modules` works now
* `make parallel`
    End of output of `make parallel`: 940 tests considered
    End of output of `make all`: 940 tests considered

objmagic committed:
Fix in testsuite: `make one` and `make parallel`

The `parallel` and `one DIR=...` targets currently do not work as
inteded for tests using the new 'ocamltest' framework: `one ...` has
no effect, and `parallel` only runs on old-style tests and ignores new
tests. This commit fixes this discrepancy in behavior.

gasche reviewed (dismissed):
I believe the patch is correct; good to go if the CI passes.

gasche commented:
Note: there was a previous attempt to fix this issue ( https://github.com/ocaml/ocaml/pull/1530 ), and it ran into trouble with the CI, so it is important to check that the CI passes indeed.

shindere commented:
I think this change is wrong.

Indeed, the way things work, it is very important that exec-one executes
only the legacy tests, because the new tests will be run later.

Or am I missing something?

gasche commented:
Ah, indeed, this is the part that was an issue.

@objmagic, I think that you need to change the `legacy*` rules to *not* run `exec-one` when `[ -f $(DIR)/ocamltests ]` holds.

objmagic commented:
@gasche @shindere 

https://github.com/ocaml/ocaml/pull/1574/commits/433309b68de785bff7a91f09e1790683a2b4eca5 makes sure that `legacy` does not run ocamltest-based tests

@shindere 

An error occasionally shows up when I test `make clean && make parallel` on my 24-core workstation:

```
Running tests from 'tests/ast-invariants' ...
mkdir: cannot create directory ‘/aaa/bbb/projects/ocaml/testsuite/_ocamltest’: File exists
Sysem command mkdir "/aaa/bbb/projects/ocaml/testsuite/_ocamltest" failed with status 1
```

I made a tentative fix https://github.com/ocaml/ocaml/pull/1574/commits/641d5192218ba4e6b7a581643714eb8898eddd30 but I am not very sure it is definitely a correct change

shindere commented:
Runhang Li (2018/01/22 19:43 +0000):
> @gasche @shindere 
> 
> https://github.com/ocaml/ocaml/pull/1574/commits/433309b68de785bff7a91f09e1790683a2b4eca5 makes sure that `legacy` does not run ocamltest-based tests
> 
> @shindere 
> 
> An error occasionally shows up when I test `make clean && make parallel` on my 24-core workstation:
> 
> ```
> Running tests from 'tests/ast-invariants' ...
> mkdir: cannot create directory ‘/aaa/bbb/projects/ocaml/testsuite/_ocamltest’: File exists
> Sysem command mkdir "/aaa/bbb/projects/ocaml/testsuite/_ocamltest" failed with status 1
> ```
> 
> I made a tentative fix
> https://github.com/ocaml/ocaml/pull/1574/commits/641d5192218ba4e6b7a581643714eb8898eddd30 but I am not very sure it is definitely a correct change

Wel mkdir -p is not portable so won't work on Windows.

There is a plan to add mkdir and rmdir to OCaml's stdlib, that's on my
todolist.

On top of that we will be able to implement mkdir -p in OCaml, I think
that's the long-term, clean way.

objmagic commented:
@shindere I see. I just removed the mkdir change.

objmagic commented:
@gasche is my change good to go?

gasche commented:
I tested the proposed change on the targets `all, legacy, new, parallel, one`.

For the first three (`all`, `legacy`, `new`), the behavior is the same as in trunk -- according to `report` at least. For `parallel`, trunk is broken (only "legacy" tests are run) and the branch works correctly (all tests are run). For `one`, trunk is broken (new tests are not run) and the branch works correctly.

gasche commented:
On the other hand, I am a bit worried by the `mkdir` race condition worried above. I wonder if it is possible to invoke the tool so that no shared `testsuite/_ocamltest` directory is created at all, and `_ocamltest` directories are created from the test directories?

gasche committed:
testsuite/Makefile: fix _ocamltest race with per-test OCAMLTESTDIR setting

When using 'make parallel', one could sometimes observe an error due
to a race condition in creating the global _ocamltest repository used
by all new-style (ocamltest) tests. This commit removes the race by
making exec-one use local _ocamltest directories, one for each test
directory (the parallelism of the 'parallel' target is at the
granularity of each test directory, not each test file, so this
is safe)

On the contrary, the global 'new' and 'new-without-report' targets
still run all test in a global BASEDIR/_ocamltest directory. This
choice was done to minimize difference in behavior for users of these
targets, but it could be revisited in the future.

gasche committed:
Changes entry

gasche commented:
I pushed an extra commit, https://github.com/ocaml/ocaml/pull/1574/commits/5130e8a3ace69788f8313cb68231c3b30b7e6d59 , that fixes the race condition by setting `OCAMLTESTDIR` to each local test directory in the `exec-one` target (the `new` target is unchanged and keeps a global `_ocamltest` directory for all tests).

objmagic commented:
Thanks @gasche! Let me test `make parallel` out on my machine tomorrow

objmagic committed:
Better changelog entry

gasche commented:
@objmagic: please feel free to squash together the two "Change entry" commits.

objmagic commented:
`make parallel` works on my workstation now - no race condition in ocamltest.

@gasche do we need to? I thought all commits will be squashed into one single commit when you press merge button.

shindere commented:
Runhang Li (2018/01/30 16:55 -0800):
> @gasche do we need to? I thought all commits will be squashed into one
> single commit when you press merge button.

Merge can be done in different ways, either squahsing or not squashing
the commits. Some times it's better not to squash them, when they convey
useful information.
